api/src/main/java/org/openmrs/CohortMembership.java

```diff
@@ -99,6 +118,18 @@ public void setEndDate(Date endDate) {
 
 	@Override
 	public int compareTo(CohortMembership o) {
-		return this.getPatient().getPatientId() - o.getPatient().getPatientId();
+		if (Objects.equals(this.getPatient().getPatientId(), o.getPatient().getPatientId())
+				&& Objects.equals(this.getCohort().getCohortId(), o.getCohort().getCohortId())
+				&& this.getStartDate().equals(o.getStartDate())
+				&& ((this.getEndDate() != null && o.getEndDate() != null
+				&& this.getEndDate().equals(o.getEndDate()))
+				|| (this.getEndDate() == null && o.getEndDate() == null))) {
+			return 0;
+		} else if ((this.isMembershipActive() && !o.isMembershipActive())
+			|| (!this.isMembershipActive() && o.isMembershipActive())) {
```

Comment:
compareTo is what gets called  when sorting since you are using a TreeSet, meaning that if you implement it this way, the voided memberships will automatically get pushed to the end, so you don't need to implement it anywhere else